Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2IdleHandler.swift

```diff
@@ -0,0 +1,170 @@
+        context.fireChannelActive()
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = self.unwrapInboundIn(data)
+
+        switch frame.payload {
+        case .goAway:
+            let action = self.state.goAwayReceived()
+            self.run(action, context: context)
+        case .settings(.settings(let settings)):
+            let action = self.state.settingsReceived(settings)
+            self.run(action, context: context)
+        default:
+            // We're not interested in other events.
+            ()
+        }
+
+        context.fireChannelRead(data)
+    }
+
+    func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+        switch event {
+        case is NIOHTTP2StreamCreatedEvent:
+            let action = self.state.streamCreated()
+            self.run(action, context: context)
+            context.fireUserInboundEventTriggered(event)
+        case is NIOHTTP2.StreamClosedEvent:
+            let action = self.state.streamClosed()
+            self.run(action, context: context)
+            context.fireUserInboundEventTriggered(event)
+        default:
+            context.fireUserInboundEventTriggered(event)
+        }
+    }
+
+    func run(_ action: StateMachine.Action, context: ChannelHandlerContext) {
+        switch action {
+        case .nothing:
+            break
+        case .notifyConnectionNewSettings(let settings):
+            self.connection.http2SettingsReceived(settings)
+        case .notifyConnectionStreamClosed(let currentlyAvailable):
+            self.connection.http2StreamClosed(availableStreams: currentlyAvailable)
+        case .notifyConnectionGoAwayReceived:
+            self.connection.http2GoAwayReceived()
+        }
+    }
+}
+
+extension HTTP2IdleHandler {
+    struct StateMachine {
+        enum Action {
+            case notifyConnectionNewSettings(HTTP2Settings)
+            case notifyConnectionGoAwayReceived
+            case notifyConnectionStreamClosed(currentlyAvailable: Int)
+            case nothing
+        }
+
+        enum State {
+            case initialized
+            case connected
+            case active(openStreams: Int, maxStreams: Int)
+            case goAwayReceived(openStreams: Int, maxStreams: Int)
+            case closed
+        }
+
+        var state: State = .initialized
+
+        mutating func connected() {
+            guard case .initialized = self.state else {
+                preconditionFailure("Invalid state")
+            }
+
+            self.state = .connected
+        }
+
+        mutating func settingsReceived(_ settings: HTTP2Settings) -> Action {
+            guard case .connected = self.state else {
+                preconditionFailure("Invalid state")
+            }
+
+            let maxStream = settings.first(where: { $0.parameter == .maxConcurrentStreams })?.value ?? 100
+
+            self.state = .active(openStreams: 0, maxStreams: maxStream)
```

Comment:
> A SETTINGS frame MUST be sent by both endpoints at the start of a connection and MAY be sent at any other time by either endpoint over the lifetime of the connection.

We need to handle this otherwise we'll blow away our open stream count. Moreover we can't always assume 100 as the default when no setting is present (since a mid-connection settings update may change another setting and we don't want to inadvertently overwrite the previous value of max concurrent streams).